File: progs/kaidoku-skeleton.py
# kaidoku.py
# 解読プログラム
# 入力: 暗号文の文字列
# 出力: 復号した平文

import os

### Global variables
ALPHABET = range(ord('a'), ord('z')+1) # 英字小文字アルファベット

# hukuho.py から
# dec(秘密鍵 k, 暗号文 c) = 平文
def dec(k, c):
  global ALPHABET
  cipher = list(c.encode("ascii"))       # 文字列 -> 文字コードの配列
  plain = cipher.copy()                  # 平文格納用配列
  for i,code in enumerate(cipher):
    if code in ALPHABET:                 # 小文字アルファベットならば
       #
       # ここに埋め込む
       #
    else:
      plain[i] = code
  return(bytes(plain).decode("ascii"))

def correlation (ciphertext):
  global ALPHABET
  length = len(ciphertext)
  ct  = list(ciphertext.encode("ascii"))
  freq = [0]*len(ALPHABET)
  for i, moji in enumerate(ct): # 頻度配列の作成
    code = ct[i]                #   i 文字目のコードを得る
    offset = code - ALPHABET[0] #   文字 a との差分
    #
    # ここに埋め込む
    #
  max = 0               # 最頻文字の位置（差分）の確定
  maxj = None
  for j in range(len(ALPHABET)):
    if freq[j] > max:
      max =  # ここに埋め込む
      maxj = # ここに埋め込む
  return( # ここに埋め込む ) # 'e' との差

def cipher_breaking (ciphertext):
  keys=correlation(ciphertext)
  if keys is None:
    raise ValueError("Cannot find a max value")
  return("key:"+str(keys)+"\ntext:"+dec(keys,ciphertext))

### TEST HARNESS
os.system("clear")
print(cipher_breaking(input()))

File: progs/kaidoku-skeleton.rb
# kaidoku.rb
# 解読プログラム
# 入力: 暗号文の文字列
# 出力: 復号した平文

# 平文を暗号化するサブルーチン
# dec(秘密鍵 k, 暗号文 c) = 平文 m
def dec(k, c)
  code_a = 97                # 文字 a の文字コード
  kosu = 26                  # 英字アルファベットの数

  leng = c.length            # 文字列の長さ
  a = c.unpack("C*")         # 文字列から文字コードの配列へ変換
  b = Array.new(leng)        # 暗号文（のコード）格納用配列

  for i in 0..leng-1
    code = a[i]                 # i 文字目のコードを得る
    sa = code - code_a          # 文字 a からの差分
    #
    # ここに埋め込む
    #
    b[i] = sa + code_a
  end
  m = b.pack("C*")              # コードの配列を文字列に直す
  return(m)
end
# サブルーチン dec （終）

##### プログラム本文 #####
# 入力と準備
code_a = 97                # 文字 a の文字コード
kosu = 26                  # 英字アルファベットの数
angobun = gets.chomp       # 暗号文を入力（注：入力文字列から改行を除去）
leng = angobun.length      # 暗号文の長さを leng に格納
a = angobun.unpack("C*")   # 文字列から文字コードの配列へ変換
hindo = Array.new(kosu, 0) # 頻度配列の準備

# 頻度配列の作成
for i in 0..leng-1
  code = a[i]              # i 文字目のコードを得る
  sa = code - code_a       # 文字 a からの差分
  #
  # ここに埋め込む
  #
end
puts(hindo)  ## 確認用

# 最頻文字の位置（差分）の確定
max = 0
maxj = -1
for j in 0..(kosu-1)
  if 
    max = 
    maxj = 
  end
end
puts(max, maxj)  ## 確認用

k =                        # maxj から暗号鍵を求める
hirabun = dec(k, angobun)  # この暗号鍵で平文に変換
puts(hirabun)              # 解読平文を出力
